framework/packages/abstract-std/src/objects/ownership/gov_ownable.rs

```diff
@@ -0,0 +1,674 @@
+        if *addr == owner {
+            return Ok(true);
+        }
+    }
+
+    Ok(false)
+}
+
+/// Assert that an account is the contract's current owner.
+pub fn assert_owner(
+    store: &dyn Storage,
+    querier: &QuerierWrapper,
+    sender: &Addr,
+) -> Result<(), GovOwnershipError> {
+    let ownership = OWNERSHIP.load(store)?;
+    // If current sender is owner of this account - it's the owner
+    if check_owner(querier, &ownership, sender).is_ok() {
+        return Ok(());
+    }
+    // Otherwise we need to check top level owner
+    let top_level_ownership = if let GovernanceDetails::SubAccount { manager, .. } = ownership.owner
+    {
+        query_top_level_owner(querier, manager)?
+    } else {
+        ownership
+    };
+    // the contract must have an owner
+    check_owner(querier, &top_level_ownership, sender)
+}
+
+/// Assert that an account is the contract's current owner.
+fn check_owner(
+    querier: &QuerierWrapper,
+    ownership: &Ownership<Addr>,
+    sender: &Addr,
+) -> Result<(), GovOwnershipError> {
+    // the contract must have an owner
+    let Some(current_owner) = &ownership.owner.owner_address(querier) else {
+        return Err(GovOwnershipError::NoOwner);
+    };
+
+    // the sender must be the current owner
+    if sender != current_owner {
+        return Err(GovOwnershipError::NotOwner);
+    }
+
+    Ok(())
+}
+
+/// Update the contract's ownership info based on the given action.
+/// Return the updated ownership.
+pub fn update_ownership(
+    deps: DepsMut,
+    block: &BlockInfo,
+    sender: &Addr,
+    version_control: Addr,
+    action: GovAction,
+) -> Result<Ownership<Addr>, GovOwnershipError> {
+    match action {
+        GovAction::TransferOwnership { new_owner, expiry } => {
+            transfer_ownership(deps, sender, new_owner, version_control, expiry)
+        }
+        GovAction::AcceptOwnership => accept_ownership(deps.storage, &deps.querier, block, sender),
+        GovAction::RenounceOwnership => renounce_ownership(deps.storage, &deps.querier, sender),
+    }
+}
+
+/// Get the current ownership value.
+pub fn get_ownership(storage: &dyn Storage) -> StdResult<Ownership<Addr>> {
+    OWNERSHIP.load(storage)
+}
+
+pub fn query_ownership<Q: CustomQuery>(
+    querier: &QuerierWrapper<Q>,
+    remote_contract: Addr,
+) -> StdResult<Ownership<Addr>> {
+    OWNERSHIP.query(querier, remote_contract)
+}
+
+/// Propose to transfer the contract's ownership to the given address, with an
+/// optional deadline.
+fn transfer_ownership(
+    deps: DepsMut,
+    sender: &Addr,
+    new_owner: GovernanceDetails<String>,
+    version_control: Addr,
+    expiry: Option<Expiration>,
+) -> Result<Ownership<Addr>, GovOwnershipError> {
+    let new_owner = new_owner.verify(deps.as_ref(), version_control)?;
+
+    if let GovernanceDetails::Renounced {} = new_owner {
+        return Err(GovOwnershipError::TransferToRenounced {});
+    }
+
+    OWNERSHIP.update(deps.storage, |ownership| {
```

Comment:
How are we handeling transfers from a regular account to a sub-account? From this it seems like you can propose it but then somehow the manager of the top-level account must accept the transfer. Is this the expected behaviour? I thought we had a different system before where we asserted that the top-level owner of the *proposed* sub-account was the sender (this ensured that one can only add sub-accounts to accounts you already own.)

Reply:
Well noted! Accidentally removed option of transferring account not as top level owner, just fixed it and covered it with tests, that don't use recursive querying :)

Reply:
Moving check of pending direct/top level into gov-ownable as well from manager right now

Reply:
Done!